mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,803 @@
+    for FetchingTransactionsData<Coin, Storage>
+{
+}
+
+impl<Coin, Storage> TransitionFrom<FetchingTransactionsData<Coin, Storage>>
+    for WaitForHistoryUpdateTrigger<Coin, Storage>
+{
+}
+
+#[async_trait]
+impl<Coin, Storage> State for WaitForHistoryUpdateTrigger<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        loop {
+            Timer::sleep(30.).await;
+
+            let ctx_balances = ctx.balances.clone();
+
+            let balances = match ctx.coin.all_balances().await {
+                Ok(balances) => balances,
+                Err(_) => {
+                    return Self::change_state(OnIoErrorCooldown::new(
+                        self.address.clone(),
+                        self.last_height_state,
+                        self.tendermint_assets_conf,
+                    ));
+                },
+            };
+
+            if balances != ctx_balances {
+                // Update balances
+                ctx.balances = balances;
+
+                return Self::change_state(FetchingTransactionsData::new(
+                    self.address.clone(),
+                    self.last_height_state,
+                    self.tendermint_assets_conf,
+                ));
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl<Coin, Storage> State for FetchingTransactionsData<Coin, Storage>
+where
+    Coin: TendermintTxHistoryOps,
+    Storage: TxHistoryStorage,
+{
+    type Ctx = TendermintTxHistoryCtx<Coin, Storage>;
+    type Result = ();
+
+    async fn on_changed(self: Box<Self>, ctx: &mut Self::Ctx) -> StateResult<Self::Ctx, Self::Result> {
+        const TX_PAGE_SIZE: u8 = 50;
+        const DEFAULT_DECIMALS: u8 = 6;
+        const DEFAULT_TRANSFER_EVENT_COUNT: usize = 2;
+        const AMOUNT_TAG_KEY: &str = "amount";
+        const RECEIVER_TAG_KEY: &str = "receiver";
+        const SPENDER_TAG_KEY: &str = "spender";
+        const COIN_RECEIVED_EVENT: &str = "coin_received";
+        const COIN_SPENT_EVENT: &str = "coin_spent";
+
+        struct TxAmounts {
+            total: BigDecimal,
+            spent_by_me: BigDecimal,
+            received_by_me: BigDecimal,
+            my_balance_change: BigDecimal,
+        }
+
+        fn get_decimals_by_denom(tendermint_assets_conf: &[Json], denom: &str) -> Option<u8> {
+            let coin = tendermint_assets_conf.iter().find(|coin| {
+                coin["protocol"]["protocol_data"]["denom"]
+                    .as_str()
+                    // just to be sure for case sensivity
+                    .map(|t| t.to_lowercase())
+                    == Some(denom.to_lowercase())
+            })?;
+
+            coin["protocol"]["protocol_data"]["decimals"]
+                .as_u64()
+                .map(|d| d.into_or(DEFAULT_DECIMALS))
```

Comment:
We can get the history record for an asset, which is not added yet into coins conf, but has decimals != 6. If the asset is then added, we will display an invalid value in history. As you mentioned in https://github.com/KomodoPlatform/atomicDEX-API/pull/1526#discussion_r1031891461, remapping is an overhead and complication of the flow. Though, searching for decimals by denom also adds some complexity, and might work incorrectly in some cases.

Reply:
> We can get the history record for an asset, which is not added yet into coins conf, but has decimals != 6. If the asset is then added, we will display an invalid value in history.

Unsupported assets are skipped here https://github.com/KomodoPlatform/atomicDEX-API/blob/f0049e9d0f25b64c43c2b6a8794059dea25b7309/mm2src/coins/tendermint/tendermint_tx_history_v2.rs#L513-L522

Decimals can only be wrong if coins file doesn't have decimals field.

I guess I have to insert all the assets even if they are not supported.

Reply:
> I guess I have to insert all the assets even if they are not supported.

Yeah, considering https://github.com/KomodoPlatform/atomicDEX-API/pull/1526#discussion_r1033960608, if we skip transactions from an old block, we will never have them in history until DB is cleaned up for some reason.